explorer/common/arena_test.cpp

```diff
@@ -0,0 +1,78 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "explorer/common/arena.h"
+
+#include <gtest/gtest.h>
+
+namespace Carbon {
+
+class ReportDestruction {
+ public:
+  explicit ReportDestruction(bool* destroyed) : destroyed_(destroyed) {}
+
+  ~ReportDestruction() { *destroyed_ = true; }
+
+ private:
+  bool* destroyed_;
+};
+
+TEST(ArenaTest, BasicAllocation) {
+  bool destroyed = false;
+  {
+    Arena arena;
+    (void)arena.New<ReportDestruction>(&destroyed);
+  }
+  EXPECT_TRUE(destroyed);
+}
+
+struct CanonicalizedDummy {
+  explicit CanonicalizedDummy(int) {}
+  explicit CanonicalizedDummy(int*) {}
+  explicit CanonicalizedDummy(int, int*) {}
+  using EnableCanonicalizedAllocation = void;
+};
+
+TEST(ArenaTest, Canonicalization) {
```

Comment:
One last thought on these tests, you might consider std::vector and std::nullopt variants to consider that functionality, since they're special-cased in Arena code.